packages/contracts-bedrock/tasks/check-l2.ts

```diff
@@ -0,0 +1,628 @@
+    }
+
+    if (addr === predeploys.GovernanceToken || addr === predeploys.ProxyAdmin) {
+      continue
+    }
+
+    const slot = await hre.ethers.provider.getStorageAt(addr, adminSlot)
+    const admin = hre.ethers.utils.hexConcat([
+      '0x000000000000000000000000',
+      predeploys.ProxyAdmin,
+    ])
+
+    if (admin !== slot) {
+      throw new Error(`incorrect admin slot in ${addr}`)
+    }
+  }
+}
+
+// assertSemver will ensure that the semver is the correct version
+const assertSemver = (version: string, name: string, override?: string) => {
+  let target = expectedSemver
+  if (override) {
+    target = override
+  }
+  if (version !== target) {
+    throw new Error(
+      `${name}: version mismatch. Got ${version}, expected ${target}`
+    )
+  }
+  console.log(`  - version: ${version}`)
+}
+
+// checkProxy will print out the proxy slots
+const checkProxy = async (hre: HardhatRuntimeEnvironment, name: string) => {
+  const address = predeploys[name]
+  if (!address) {
+    throw new Error(`unknown contract name: ${name}`)
+  }
+
+  const impl = await hre.ethers.provider.getStorageAt(address, implSlot)
+  const admin = await hre.ethers.provider.getStorageAt(address, adminSlot)
+
+  console.log(`  - EIP-1967 implementation slot: ${impl}`)
+  console.log(`  - EIP-1967 admin slot: ${admin}`)
+}
+
+// assertProxy will require the proxy is set
+const assertProxy = async (hre: HardhatRuntimeEnvironment, name: string) => {
+  const address = predeploys[name]
+  if (!address) {
+    throw new Error(`unknown contract name: ${name}`)
+  }
+
+  const code = await hre.ethers.provider.getCode(address)
+  const deployInfo = await hre.artifacts.readArtifact('Proxy')
+
+  if (code !== deployInfo.deployedBytecode) {
+    throw new Error(`${address}: code mismatch`)
+  }
+
+  const impl = await hre.ethers.provider.getStorageAt(address, implSlot)
+  const implAddress = '0x' + impl.slice(26)
+  const implCode = await hre.ethers.provider.getCode(implAddress)
+  if (implCode === '0x') {
+    throw new Error('No code at implementation')
+  }
+}
+
+const check = {
+  // LegacyMessagePasser
+  // - check version
+  // - is behind a proxy
+  LegacyMessagePasser: async (hre: HardhatRuntimeEnvironment) => {
+    const LegacyMessagePasser = await hre.ethers.getContractAt(
+      'LegacyMessagePasser',
+      predeploys.LegacyMessagePasser
+    )
+
+    const version = await LegacyMessagePasser.version()
+    assertSemver(version, 'LegacyMessagePasser')
+
+    await checkProxy(hre, 'LegacyMessagePasser')
+    await assertProxy(hre, 'LegacyMessagePasser')
+  },
+  // DeployerWhitelist
+  // - check version
+  // - is behind a proxy
+  // - owner is `address(0)`
+  DeployerWhitelist: async (hre: HardhatRuntimeEnvironment) => {
+    const DeployerWhitelist = await hre.ethers.getContractAt(
+      'DeployerWhitelist',
+      predeploys.DeployerWhitelist
+    )
+
+    const version = await DeployerWhitelist.version()
+    assertSemver(version, 'DeployerWhitelist')
```

Comment:
Nitpick but you could just pass the contract instance in here to avoid having to do `await contract.version()` every time